File: netscope-go/internal/model/json.go
package model

import (
	"sort"
	"strconv"
	"strings"
	"time"
)

// Canonical JSON emitter.
//
// This is hand-rolled rather than encoding/json because the output must be
// BYTE-IDENTICAL to the C++ implementation's emitter so that a parity run can
// diff the two (spec §9). encoding/json's float formatting (shortest
// round-trip) is not reproducible in C++, so every float goes out with exactly
// three decimals and every field has a fixed position.
//
// Rules, mirrored in netscope-cpp/src/core/json.cpp:
//   - 2-space indent, one field or array element per line
//   - object fields emitted in the order written below, never sorted
//   - floats: strconv 'f' with precision 3; -0 normalized to 0
//   - durations: integer milliseconds (truncated)
//   - nil optional -> null
//   - empty array -> []

type jsonWriter struct {
	b      strings.Builder
	indent int
	// first tracks whether the next member of the current container is the first,
	// so we know when to emit a separating comma.
	first []bool
}

func newJSONWriter() *jsonWriter {
	return &jsonWriter{first: []bool{true}}
}

func (w *jsonWriter) pad() {
	for i := 0; i < w.indent; i++ {
		w.b.WriteString("  ")
	}
}

func (w *jsonWriter) sep() {
	if w.first[len(w.first)-1] {
		w.first[len(w.first)-1] = false
	} else {
		w.b.WriteString(",\n")
		return
	}
	w.b.WriteString("\n")
}

func (w *jsonWriter) push() {
	w.indent++
	w.first = append(w.first, true)
}

func (w *jsonWriter) pop() {
	w.indent--
	w.first = w.first[:len(w.first)-1]
	w.b.WriteString("\n")
	w.pad()
}

func (w *jsonWriter) beginObject(key string) {
	w.sep()
	w.pad()
	if key != "" {
		w.b.WriteString(quoteJSON(key))
		w.b.WriteString(": ")
	}
	w.b.WriteString("{")
	w.push()
}

func (w *jsonWriter) endObject() {
	if w.first[len(w.first)-1] {
		// empty object: no newline padding
		w.indent--
		w.first = w.first[:len(w.first)-1]
		w.b.WriteString("}")
		return
	}
	w.pop()
	w.b.WriteString("}")
}

func (w *jsonWriter) beginArray(key string) {
	w.sep()
	w.pad()
	w.b.WriteString(quoteJSON(key))
	w.b.WriteString(": [")
	w.push()
}

func (w *jsonWriter) endArray() {
	if w.first[len(w.first)-1] {
		w.indent--
		w.first = w.first[:len(w.first)-1]
		w.b.WriteString("]")
		return
	}
	w.pop()
	w.b.WriteString("]")
}

func (w *jsonWriter) raw(key, val string) {
	w.sep()
	w.pad()
	w.b.WriteString(quoteJSON(key))
	w.b.WriteString(": ")
	w.b.WriteString(val)
}

func (w *jsonWriter) str(key, val string)        { w.raw(key, quoteJSON(val)) }
func (w *jsonWriter) int64f(key string, v int64) { w.raw(key, strconv.FormatInt(v, 10)) }
func (w *jsonWriter) intf(key string, v int)     { w.raw(key, strconv.Itoa(v)) }
func (w *jsonWriter) uintf(key string, v uint64) {
	w.raw(key, strconv.FormatUint(v, 10))
}
func (w *jsonWriter) boolf(key string, v bool) {
	if v {
		w.raw(key, "true")
	} else {
		w.raw(key, "false")
	}
}
func (w *jsonWriter) msf(key string, d time.Duration) {
	w.int64f(key, d.Milliseconds())
}
func (w *jsonWriter) optf(key string, v *float64) {
	if v == nil {
		w.raw(key, "null")
		return
	}
	w.raw(key, FormatF3(*v))
}
func (w *jsonWriter) optInt(key string, v *int) {
	if v == nil {
		w.raw(key, "null")
		return
	}
	w.intf(key, *v)
}

// FormatF3 renders a float with exactly three decimals, normalizing -0.
// Both implementations must agree on this exact rendering.
func FormatF3(v float64) string {
	s := strconv.FormatFloat(v, 'f', 3, 64)
	if s == "-0.000" {
		return "0.000"
	}
	return s
}

func quoteJSON(s string) string {
	var b strings.Builder
	b.WriteByte('"')
	for i := 0; i < len(s); i++ {
		c := s[i]
		switch c {
		case '"':
			b.WriteString(`\"`)
		case '\\':
			b.WriteString(`\\`)
		case '\n':
			b.WriteString(`\n`)
		case '\r':
			b.WriteString(`\r`)
		case '\t':
			b.WriteString(`\t`)
		default:
			if c < 0x20 {
				b.WriteString(`\u00`)
				const hex = "0123456789abcdef"
				b.WriteByte(hex[c>>4])
				b.WriteByte(hex[c&0xf])
			} else {
				b.WriteByte(c)
			}
		}
	}
	b.WriteByte('"')
	return b.String()
}

func (w *jsonWriter) stats(key string, s RttStats) {
	w.beginObject(key)
	w.intf("samples", s.Samples)
	w.optf("lastMs", s.LastMs)
	w.optf("bestMs", s.BestMs)
	w.optf("avgMs", s.AvgMs)
	w.optf("worstMs", s.WorstMs)
	w.optf("jitterMs", s.JitterMs)
	w.optf("stdevMs", s.StdevMs)
	w.beginArray("spark")
	for _, v := range s.Spark {
		w.sep()
		w.pad()
		w.b.WriteString(FormatF3(v))
	}
	w.endArray()
	w.endObject()
}

// CanonicalJSON serializes a snapshot for parity comparison.
func (s *Snapshot) CanonicalJSON() string {
	w := newJSONWriter()
	// top-level object, written without the leading separator logic
	w.b.WriteString("{")
	w.push()

	w.uintf("revision", s.Revision)
	w.uintf("generation", s.Generation)
	w.str("mode", string(s.Mode))
	w.boolf("degraded", s.Degraded)
	w.boolf("paused", s.Paused)
	w.msf("nowMs", s.Now)

	w.beginObject("target")
	w.str("input", s.Target.Input)
	w.str("ip", s.Target.IP)
	w.str("family", string(s.Target.Family))
	w.msf("resolvedAtMs", s.Target.ResolvedAt)
	w.endObject()

	w.beginObject("cadence")
	w.intf("destIntervalMs", s.Cadence.DestIntervalMs)
	w.intf("midIntervalMs", s.Cadence.MidIntervalMs)
	w.intf("globalCapPps", s.Cadence.GlobalCapPPS)
	w.intf("windowDurationMs", s.Cadence.WindowDurationMs)
	w.intf("probeTimeoutMs", s.Cadence.ProbeTimeoutMs)
	w.endObject()

	w.beginArray("hops")
	for i := range s.Hops {
		h := &s.Hops[i]
		w.beginObject("")
		w.intf("ttl", h.TTL)
		w.str("status", string(h.Status))
		w.uintf("sent", h.Sent)
		w.uintf("replied", h.Replied)
		w.optf("lossPct", h.LossPct)
		w.str("primary", h.Primary)
		w.boolf("isDestination", h.IsDestination)
		w.stats("stats", h.Stats)
		w.beginArray("responders")
		for j := range h.Responders {
			r := &h.Responders[j]
			w.beginObject("")
			w.str("ip", r.IP)
			w.str("rdns", r.RDNS)
			w.str("asn", r.ASN)
			w.str("org", r.Org)
			w.uintf("seen", r.Seen)
			w.msf("firstSeenAtMs", r.FirstSeenAt)
			w.msf("lastSeenAtMs", r.LastSeenAt)
			w.stats("stats", r.Stats)
			w.endObject()
		}
		w.endArray()
		w.endObject()
	}
	w.endArray()

	w.beginObject("local")
	w.str("interface", s.Local.Interface)
	w.str("address", s.Local.Address)
	w.str("gateway", s.Local.Gateway)
	w.beginArray("dnsServers")
	dns := append([]string(nil), s.Local.DNSServers...)
	sort.Strings(dns)
	for _, d := range dns {
		w.sep()
		w.pad()
		w.b.WriteString(quoteJSON(d))
	}
	w.endArray()
	w.str("defaultRoute", s.Local.DefaultRoute)
	w.str("publicIp", s.Local.PublicIP)
	w.str("note", s.Local.Note)
	w.endObject()

	w.beginObject("health")
	w.intf("httpStatus", s.Health.HTTPStatus)
	w.optf("httpLatencyMs", s.Health.HTTPLatency)
	w.str("httpNote", s.Health.HTTPNote)
	w.intf("tcpPort", s.Health.TCPPort)
	w.boolf("tcpOpen", s.Health.TCPOpen)
	w.str("tcpNote", s.Health.TCPNote)
	w.endObject()

	w.beginArray("events")
	for i := range s.Events {
		e := &s.Events[i]
		w.beginObject("")
		w.msf("atMs", e.At)
		w.str("kind", string(e.Kind))
		w.optInt("ttl", e.TTL)
		w.str("text", e.Text)
		w.endObject()
	}
	w.endArray()

	w.pop()
	w.b.WriteString("}\n")
	return w.b.String()
}

File: netscope-cpp/src/json.cpp
#include "json.h"

#include <algorithm>
#include <array>
#include <cctype>
#include <cstdio>
#include <cstdlib>

namespace netscope {
namespace {

// Mirrors the Go writer's state machine exactly: 2-space indent, one field or
// element per line, a leading "\n" before the first member and ",\n" before the
// rest.
class Writer {
public:
    std::string out;

    void openTop() {
        out += '{';
        push();
    }
    void closeTop() {
        pop();
        out += "}\n";
    }

    void beginObject(std::string_view key) {
        sep();
        padOut();
        if (!key.empty()) {
            quote(key);
            out += ": ";
        }
        out += '{';
        push();
    }
    void endObject() {
        if (first_.back()) {
            --indent_;
            first_.pop_back();
            out += '}';
            return;
        }
        pop();
        out += '}';
    }

    void beginArray(std::string_view key) {
        sep();
        padOut();
        quote(key);
        out += ": [";
        push();
    }
    void endArray() {
        if (first_.back()) {
            --indent_;
            first_.pop_back();
            out += ']';
            return;
        }
        pop();
        out += ']';
    }

    void raw(std::string_view key, std::string_view value) {
        sep();
        padOut();
        quote(key);
        out += ": ";
        out += value;
    }
    void str(std::string_view key, std::string_view value) {
        sep();
        padOut();
        quote(key);
        out += ": ";
        quote(value);
    }
    void boolean(std::string_view key, bool v) { raw(key, v ? "true" : "false"); }

    void integer(std::string_view key, long long v) { raw(key, std::to_string(v)); }
    void unsignedInt(std::string_view key, unsigned long long v) { raw(key, std::to_string(v)); }

    void millis(std::string_view key, Dur d) {
        integer(key, std::chrono::duration_cast<std::chrono::milliseconds>(d).count());
    }

    void optDouble(std::string_view key, const std::optional<double>& v) {
        if (!v) {
            raw(key, "null");
            return;
        }
        raw(key, formatF3(*v));
    }
    void optInt(std::string_view key, const std::optional<int>& v) {
        if (!v) {
            raw(key, "null");
            return;
        }
        integer(key, *v);
    }

    // Bare array element (no key).
    void element(std::string_view literal) {
        sep();
        padOut();
        out += literal;
    }
    void elementString(std::string_view value) {
        sep();
        padOut();
        quote(value);
    }
    void beginElementObject() {
        sep();
        padOut();
        out += '{';
        push();
    }

    void quote(std::string_view s) {
        out += '"';
        for (unsigned char c : s) {
            switch (c) {
                case '"': out += "\\\""; break;
                case '\\': out += "\\\\"; break;
                case '\n': out += "\\n"; break;
                case '\r': out += "\\r"; break;
                case '\t': out += "\\t"; break;
                default:
                    if (c < 0x20) {
                        static constexpr char hex[] = "0123456789abcdef";
                        out += "\\u00";
                        out += hex[c >> 4];
                        out += hex[c & 0xf];
                    } else {
                        out += static_cast<char>(c);
                    }
            }
        }
        out += '"';
    }

private:
    int indent_ = 0;
    std::vector<bool> first_{true};

    void push() {
        ++indent_;
        first_.push_back(true);
    }
    void pop() {
        --indent_;
        first_.pop_back();
        out += '\n';
        padOut();
    }
    void sep() {
        if (first_.back()) {
            first_.back() = false;
            out += '\n';
        } else {
            out += ",\n";
        }
    }
    void padOut() {
        for (int i = 0; i < indent_; ++i) out += "  ";
    }
};

void writeStats(Writer& w, std::string_view key, const RttStats& s) {
    w.beginObject(key);
    w.integer("samples", s.samples);
    w.optDouble("lastMs", s.lastMs);
    w.optDouble("bestMs", s.bestMs);
    w.optDouble("avgMs", s.avgMs);
    w.optDouble("worstMs", s.worstMs);
    w.optDouble("jitterMs", s.jitterMs);
    w.optDouble("stdevMs", s.stdevMs);
    w.beginArray("spark");
    for (double v : s.spark) w.element(formatF3(v));
    w.endArray();
    w.endObject();
}

}  // namespace

std::string formatF3(double v) {
    std::array<char, 64> buf{};
    std::snprintf(buf.data(), buf.size(), "%.3f", v);
    std::string s(buf.data());
    if (s == "-0.000") return "0.000";
    return s;
}

std::string canonicalJson(const Snapshot& s) {
    Writer w;
    w.openTop();

    w.unsignedInt("revision", s.revision);
    w.unsignedInt("generation", s.generation);
    w.str("mode", modeName(s.mode));
    w.boolean("degraded", s.degraded);
    w.boolean("paused", s.paused);
    w.millis("nowMs", s.now);

    w.beginObject("target");
    w.str("input", s.target.input);
    w.str("ip", s.target.ip);
    w.str("family", familyName(s.target.family));
    w.millis("resolvedAtMs", s.target.resolvedAt);
    w.endObject();

    w.beginObject("cadence");
    w.integer("destIntervalMs", s.cadence.destIntervalMs);
    w.integer("midIntervalMs", s.cadence.midIntervalMs);
    w.integer("globalCapPps", s.cadence.globalCapPps);
    w.integer("windowDurationMs", s.cadence.windowDurationMs);
    w.integer("probeTimeoutMs", s.cadence.probeTimeoutMs);
    w.endObject();

    w.beginArray("hops");
    for (const auto& h : s.hops) {
        w.beginElementObject();
        w.integer("ttl", h.ttl);
        w.str("status", statusName(h.status));
        w.unsignedInt("sent", h.sent);
        w.unsignedInt("replied", h.replied);
        w.optDouble("lossPct", h.lossPct);
        w.str("primary", h.primary);
        w.boolean("isDestination", h.isDestination);
        writeStats(w, "stats", h.stats);
        w.beginArray("responders");
        for (const auto& r : h.responders) {
            w.beginElementObject();
            w.str("ip", r.ip);
            w.str("rdns", r.rdns);
            w.str("asn", r.asn);
            w.str("org", r.org);
            w.unsignedInt("seen", r.seen);
            w.millis("firstSeenAtMs", r.firstSeenAt);
            w.millis("lastSeenAtMs", r.lastSeenAt);
            writeStats(w, "stats", r.stats);
            w.endObject();
        }
        w.endArray();
        w.endObject();
    }
    w.endArray();

    w.beginObject("local");
    w.str("interface", s.local.interfaceName);
    w.str("address", s.local.address);
    w.str("gateway", s.local.gateway);
    w.beginArray("dnsServers");
    {
        // Sorted so the two implementations agree regardless of discovery order.
        std::vector<std::string> dns = s.local.dnsServers;
        std::sort(dns.begin(), dns.end());
        for (const auto& d : dns) w.elementString(d);
    }
    w.endArray();
    w.str("defaultRoute", s.local.defaultRoute);
    w.str("publicIp", s.local.publicIp);
    w.str("note", s.local.note);
    w.endObject();

    w.beginObject("health");
    w.integer("httpStatus", s.health.httpStatus);
    w.optDouble("httpLatencyMs", s.health.httpLatencyMs);
    w.str("httpNote", s.health.httpNote);
    w.integer("tcpPort", s.health.tcpPort);
    w.boolean("tcpOpen", s.health.tcpOpen);
    w.str("tcpNote", s.health.tcpNote);
    w.endObject();

    w.beginArray("events");
    for (const auto& e : s.events) {
        w.beginElementObject();
        w.millis("atMs", e.at);
        w.str("kind", eventKindName(e.kind));
        w.optInt("ttl", e.ttl);
        w.str("text", e.text);
        w.endObject();
    }
    w.endArray();

    w.closeTop();
    return w.out;
}

// ---------------------------------------------------------------- reader

const JsonValue* JsonValue::find(std::string_view key) const {
    for (const auto& kv : object) {
        if (kv.first == key) return &kv.second;
    }
    return nullptr;
}

std::string JsonValue::strOr(std::string_view key, std::string_view fallback) const {
    const JsonValue* v = find(key);
    if (!v || v->type != Type::String) return std::string(fallback);
    return v->str;
}

double JsonValue::numOr(std::string_view key, double fallback) const {
    const JsonValue* v = find(key);
    if (!v || v->type != Type::Number) return fallback;
    return v->number;
}

long long JsonValue::intOr(std::string_view key, long long fallback) const {
    const JsonValue* v = find(key);
    if (!v || v->type != Type::Number) return fallback;
    return static_cast<long long>(v->number);
}

bool JsonValue::boolOr(std::string_view key, bool fallback) const {
    const JsonValue* v = find(key);
    if (!v || v->type != Type::Bool) return fallback;
    return v->boolean;
}

std::optional<double> JsonValue::optNum(std::string_view key) const {
    const JsonValue* v = find(key);
    if (!v || v->type != Type::Number) return std::nullopt;
    return v->number;
}

std::vector<std::string> JsonValue::stringsAt(std::string_view key) const {
    std::vector<std::string> out;
    const JsonValue* v = find(key);
    if (!v || v->type != Type::Array) return out;
    for (const auto& e : v->array) {
        if (e.type == Type::String) out.push_back(e.str);
    }
    return out;
}

namespace {

class Parser {
public:
    Parser(std::string_view text) : s_(text) {}

    bool parse(JsonValue& out) {
        skipWs();
        if (!parseValue(out)) return false;
        skipWs();
        return true;
    }

    std::string error;

private:
    std::string_view s_;
    std::size_t i_ = 0;

    void skipWs() {
        while (i_ < s_.size() &&
               (s_[i_] == ' ' || s_[i_] == '\t' || s_[i_] == '\n' || s_[i_] == '\r')) {
            ++i_;
        }
    }

    bool fail(const char* what) {
        error = std::string(what) + " at offset " + std::to_string(i_);
        return false;
    }

    bool literal(std::string_view lit) {
        if (s_.compare(i_, lit.size(), lit) != 0) return false;
        i_ += lit.size();
        return true;
    }

    bool parseValue(JsonValue& out) {
        if (i_ >= s_.size()) return fail("unexpected end of input");
        switch (s_[i_]) {
            case '{': return parseObject(out);
            case '[': return parseArray(out);
            case '"':
                out.type = JsonValue::Type::String;
                return parseString(out.str);
            case 't':
                if (!literal("true")) return fail("bad literal");
                out.type = JsonValue::Type::Bool;
                out.boolean = true;
                return true;
            case 'f':
                if (!literal("false")) return fail("bad literal");
                out.type = JsonValue::Type::Bool;
                out.boolean = false;
                return true;
            case 'n':
                if (!literal("null")) return fail("bad literal");
                out.type = JsonValue::Type::Null;
                return true;
            default: return parseNumber(out);
        }
    }

    bool parseObject(JsonValue& out) {
        out.type = JsonValue::Type::Object;
        ++i_;  // '{'
        skipWs();
        if (i_ < s_.size() && s_[i_] == '}') {
            ++i_;
            return true;
        }
        for (;;) {
            skipWs();
            std::string key;
            if (i_ >= s_.size() || s_[i_] != '"') return fail("expected object key");
            if (!parseString(key)) return false;
            skipWs();
            if (i_ >= s_.size() || s_[i_] != ':') return fail("expected ':'");
            ++i_;
            skipWs();
            JsonValue val;
            if (!parseValue(val)) return false;
            out.object.emplace_back(std::move(key), std::move(val));
            skipWs();
            if (i_ >= s_.size()) return fail("unterminated object");
            if (s_[i_] == ',') {
                ++i_;
                continue;
            }
            if (s_[i_] == '}') {
                ++i_;
                return true;
            }
            return fail("expected ',' or '}'");
        }
    }

    bool parseArray(JsonValue& out) {
        out.type = JsonValue::Type::Array;
        ++i_;  // '['
        skipWs();
        if (i_ < s_.size() && s_[i_] == ']') {
            ++i_;
            return true;
        }
        for (;;) {
            skipWs();
            JsonValue val;
            if (!parseValue(val)) return false;
            out.array.push_back(std::move(val));
            skipWs();
            if (i_ >= s_.size()) return fail("unterminated array");
            if (s_[i_] == ',') {
                ++i_;
                continue;
            }
            if (s_[i_] == ']') {
                ++i_;
                return true;
            }
            return fail("expected ',' or ']'");
        }
    }

    bool parseString(std::string& out) {
        ++i_;  // opening quote
        out.clear();
        while (i_ < s_.size()) {
            char c = s_[i_++];
            if (c == '"') return true;
            if (c != '\\') {
                out += c;
                continue;
            }
            if (i_ >= s_.size()) break;
            char esc = s_[i_++];
            switch (esc) {
                case '"': out += '"'; break;
                case '\\': out += '\\'; break;
                case '/': out += '/'; break;
                case 'b': out += '\b'; break;
                case 'f': out += '\f'; break;
                case 'n': out += '\n'; break;
                case 'r': out += '\r'; break;
                case 't': out += '\t'; break;
                case 'u': {
                    if (i_ + 4 > s_.size()) return fail("truncated \\u escape");
                    unsigned code = 0;
                    for (int k = 0; k < 4; ++k) {
                        char h = s_[i_ + static_cast<std::size_t>(k)];
                        code <<= 4;
                        if (h >= '0' && h <= '9') code |= static_cast<unsigned>(h - '0');
                        else if (h >= 'a' && h <= 'f') code |= static_cast<unsigned>(h - 'a' + 10);
                        else if (h >= 'A' && h <= 'F') code |= static_cast<unsigned>(h - 'A' + 10);
                        else return fail("bad \\u escape");
                    }
                    i_ += 4;
                    // Scenario files are ASCII/UTF-8; emit UTF-8 for the BMP range.
                    if (code < 0x80) {
                        out += static_cast<char>(code);
                    } else if (code < 0x800) {
                        out += static_cast<char>(0xC0 | (code >> 6));
                        out += static_cast<char>(0x80 | (code & 0x3F));
                    } else {
                        out += static_cast<char>(0xE0 | (code >> 12));
                        out += static_cast<char>(0x80 | ((code >> 6) & 0x3F));
                        out += static_cast<char>(0x80 | (code & 0x3F));
                    }
                    break;
                }
                default: return fail("unknown escape");
            }
        }
        return fail("unterminated string");
    }

    bool parseNumber(JsonValue& out) {
        std::size_t start = i_;
        if (i_ < s_.size() && (s_[i_] == '-' || s_[i_] == '+')) ++i_;
        bool any = false;
        while (i_ < s_.size() && std::isdigit(static_cast<unsigned char>(s_[i_]))) {
            ++i_;
            any = true;
        }
        if (i_ < s_.size() && s_[i_] == '.') {
            ++i_;
            while (i_ < s_.size() && std::isdigit(static_cast<unsigned char>(s_[i_]))) {
                ++i_;
                any = true;
            }
        }
        if (i_ < s_.size() && (s_[i_] == 'e' || s_[i_] == 'E')) {
            ++i_;
            if (i_ < s_.size() && (s_[i_] == '-' || s_[i_] == '+')) ++i_;
            while (i_ < s_.size() && std::isdigit(static_cast<unsigned char>(s_[i_]))) ++i_;
        }
        if (!any) return fail("expected a value");
        out.type = JsonValue::Type::Number;
        out.number = std::strtod(std::string(s_.substr(start, i_ - start)).c_str(), nullptr);
        return true;
    }
};

}  // namespace

std::optional<JsonValue> parseJson(std::string_view text, std::string& err) {
    Parser p(text);
    JsonValue v;
    if (!p.parse(v)) {
        err = p.error;
        return std::nullopt;
    }
    return v;
}

}  // namespace netscope
